### ISSUE TITLE
Contrast Checker should work for all text / background color pair settings for all blocks

### ISSUE THREAD

**Issue by afercia** [WordPress/gutenberg#60824]
### Description

Discovered while exploring https://github.com/WordPress/gutenberg/issues/58936#issuecomment-2060513454

Some blocks have a setting to change the _text_ and _background_ colors. The Contrast Checker should work for all these settings for all blocks. It appears it doesn't.

Note: there may be different reasons why it doesn't work for some blocks. Some guessing:
- Some blocks do apply the colors on their _inner blocks_ e.g. the Search block.
- For other blocks, it could be that is just buggy.

The Contrast Checker is an important feature that should work 'out of the box' for all color pairs that apply to text and background.

Since WordPress 6.6 is supposed to be a 'bug fixes' release, I'd like to propose to fix the Contrast Checker and make it work reliably for 6.6. Some e2e tests would be nice.

### Step-by-step reproduction instructions

- Create a new post and add a Paragraph, a Table, and a Search block.
- On each of these blocks, set the text and background colors to a color pair that is clearly below the accessible contrast ratio threshold. Try to get something like what is illustrated in the screenshot below:

![Screenshot 2024-04-17 at 15 58 29](https://github.com/WordPress/gutenberg/assets/1682452/68f4826f-8d6a-49dc-9d04-c1b446015618)

- Select the Paragraph block and observe in the settings sidebar, after the color settings, the Contrast Checker does appear. 
- Select the Table block and observe in the settings sidebar, after the color settings, the Contrast Checker does _not_ appear. Try to change colors, set new insufficient contrast to trigger again the Contrast Checker and observe it still does not appear.
- Select the Search block and observe in the settings sidebar, after the color settings, the Contrast Checker does _not_ appear. Try to change colors, set new insufficient contrast to trigger again the Contrast Checker and observe it still does not appear.

### Please confirm that you have searched existing issues in the repo.

Yes

### Please confirm that you have tested with all plugins deactivated except Gutenberg.

Yes

**Comment by afercia** [comment]
One more place where the ContrastChcker should work is the `InlineColorUI` component, that is: from the block toolbar > More > Highlight. This is part of the 'inline formatting' features and allows to 'highlight' the selected text by setting text and background colors 'inline'. Screenshot:

![Screenshot 2024-05-07 at 09 19 26](https://github.com/WordPress/gutenberg/assets/1682452/0981d959-b013-4f41-8aa8-2ef2ecedf8b5)

**Comment by Mamaduka** [comment]
It would be helpful if we could re-test this with #68799.

**Comment by Mamaduka** [comment]
@juanfra, have you tested these scenarios?

**Comment by afercia** [comment]
I had a chance to test https://github.com/WordPress/gutenberg/pull/68799 and it's great to see that fix merged, thanks a lot ❤ 

The contrast checker still doesn't work for blocks with inner blocks and for the `InlineColorUI` component though. I do understand https://github.com/WordPress/gutenberg/pull/68799 fixes the way colors are detected, which is the first step to make the Contrast checker work reliable and that's great.
However, blocks like Search, Table, and other blocks that apply the colors _to the inner blocks_ are incompatible with `ContrastChecker` because the parent block is unaware of the colors. I would say their _architectural_ design didn't take into consideereation the contrast checker. Is there any wah to make the parent block aware of the inner blocks colors?

One more place where it still doesn't work is for the `InlineColorUI` component, that is: the block toolbar > More > Highlight. 

Observe in the screenshot below:
- highlighted text and background color
- the search button
- the table cells 

![Image](https://github.com/user-attachments/assets/6a7f073f-35e8-4eef-95ac-4f27257c7c8b)

Edit:
Other special cases to mention that come into my mind are:
- The 'Social Icons' block applies the colors to the inner blocks as well, so the contrast checker doesn't work.
- The Contrast checker is disabled in the 'Cover' block after https://github.com/WordPress/gutenberg/issues/50531 / https://github.com/WordPress/gutenberg/pull/53080 to my understanding that was an interim solution but I'm not sure there has been any follow up.

**Comment by juanfra** [comment]
> @juanfra, have you tested these scenarios?

No, I haven’t tested those cases. I focused on checking scenarios where the contrast checker was supposed to display warnings but didn’t explore beyond that.
